### PR TITLE
[Winforms] Fix calculating TableLayout preferred size.

### PR DESCRIPTION
**Pull request by PreferLinux**
When calculating a TableLayout preferred size, don't redistribute any remaining space.

### CODE REVIEW

**Comment by filipnavara** [comment]
Do you have a case where the code behaves incorrectly? The redistribution is quite intentional due to the way the nested layouts work. Even though it may not be exactly what Microsoft code does it's tricky to get right and may trigger unintended consequences.

If you ask for preferred size using `PreferredSize` property or `GetPreferredSize(Size.Empty)`  you would not get any remaining space anyway. If you get passed `GetPreferredSize(new Size(int.MaxValue, int.MaxValue))` (or any value larger than 32768 iirc) it should be special-cased and I am not sure if that's done or not. When one specifies a reasonable width it's very much intentional to resize the columns to fit it and calculate the height based on it (!). Removing the redistribution from height might actually be correct but a test case would be very welcome to prove that.

**Comment by PreferLinux** [comment]
The problem I've seen is stuff not autosizing correctly. I'm not sure what exactly is happening – it is a while since I wrote most of this stuff, so I'll have to figure it out again. And I'll have to figure out a minimal test-case rather than a form in an application. But it seems that somehow it is getting a proposedSize based on a parent, and then using up all of proposedSize.Width (when it should actually be based on the width of a non-AutoSize Label, that happens to be inside a Dock.Top GroupBox inside an AutoSize'ing TableLayoutPanel with only one AutoSize column, in turn inside another AutoSize'ing TableLayoutPanel with two AutoSize columns but with the other column's contents hidden, all inside an AutoSize'ing Form).

Of note, the change I've made is already present for one of the two redistribution cases for height. I don't know if that means width should have it as well, if there is some good reason not to do it for width, or something else.

**Comment by filipnavara** [comment]
> Of note, the change I've made is already present for one of the two redistribution cases for height. I don't know if that means width should have it as well, if there is some good reason not to do it for width, or something else.

IIRC the idea is that if you get some non-zero `proposedSize` in `GetPreferredSize` you should try to auto-size into its width (ie. fit labels horizontally, if possible, etc.) and figure out some minimal height. to fit everything. However, there is a lot of edge cases that I don't quite remember off top of my head.

**Comment by PreferLinux** [comment]
Well, this change is at least partially incorrect. `RedistributePercents` needs to be called, so that an autosize table anchored `Left | Right` / `Top | Bottom` that has `Percent` rows / columns stretches with its parent. However, the same situation except with a table that has `AutoSize` instead of `Percent` columns shouldn't stretch – but it does (horizontally only), which must be a bug. (Both of these cases also uncover two bugs in `DefaultLayout`: if the parent is resized smaller, the table does not shrink (Windows shrinks each dimension until it is less than 0, then sets it to preferred size); and if after that the parent is resized larger, the size immediately increases when it shouldn't (the distances to the right and bottom can be significantly smaller than they were, but should be the same)).

The problem I was actually having is actually not very easy to reproduce. I found it only triggered when AutoScale kicked in, it depends on having a non-visible control in the first column, and both an autosize control and a docked non-autosize control in rows in a nested table in the second column. The main steps that happen are as follows (largely written to help me get a clear view of what is happening, but it might be helpful to others):
1. The outer table is laid out (due to `PerformLayout` on outer table). It is at full size.
  a. Size the columns. The first column has no visible controls, so gets a width of zero. The other column is initially sized according to the nested table's preferred size (calculated with `proposedSize.Width == 0` so gets the correct size), but gets given the full width of the table due to `available_width` redistribution.
  b. Size and position the controls. Uses the nested table's preferred size, calculated with `proposedSize == column_width`. The nested table asks for the full width, for the same reason as the previous step.
2. Lay out the inner table (due to `PerformLayout` on inner table). It is excessively wide due to above. If the second control does not autosize, the Visual Studio designer doesn't call `PerformLayout` here, so omits this step.
  a. Size the column as above. Once again, gets full table width.
  b. Size and position the controls. The docked control gets the full width irrespective of its preferred size. The other control doesn't matter.
3. Scale (due to `AutoScaleMode == Font`. The sizing above sets `bounds`, but not `explicit_bounds` (`BoundsSpecified.None`). When Scale does its thing, it use `BoundsSpecified.All`. So now `explicit_bounds` reflect the sizes above (which, unless step 2 was omitted, are too large). Only the docked non-autosize control is significantly affected.
4. Lay out the form, and redo table layouts. This uses the new bounds after scale, and if they are large, sizes incorrectly. The layouts nest in the opposite direction to above. For each control:
  a. If the control is autosize, calculate its preferred size (with `proposedSize` empty, meaning  tables never have `available_width` etc).
  b. Set the size (and location) of the control (to its preferred size, if autosize).
  c. If the control is a container, lay out the control (this is obviously limited by its size, so once again there is no `available_width` on autosize tables) and goto 4a for each control inside it.

So basically, sizes can get weird in the first part, and `ScaleControl` can make those sizes persist when they shouldn't...

Most of my stuff has `AutoScaleMode` set to None, so I'm not sure why I even had a problem...
